allms/models/abstract.py

```diff
@@ -105,27 +106,49 @@ def generate(
         if output_data_model_class:
             return self._parse_model_output(model_responses)
         return model_responses
+    
+    def _extract_json_from_response(self, model_response_data: ResponseData) -> str:
+        search_results = self._json_pattern.findall(model_response_data.response)
+        
+        if len(search_results) == 0:
+            return model_response_data.response
+        
+        return search_results[0]
+
+    def _parse_response(
+            self, 
+            model_response_data: ResponseData
+        ) -> ResponseWithError:
+        raw_response = self._extract_json_from_response(model_response_data)
 
-    def _parse_response(self, model_response_data: ResponseData) -> typing.Tuple[str, typing.Optional[str]]:
         try:
-            return self._parser.parse(model_response_data.response), None
+            return ResponseWithError(
+                response=self._parser.parse(raw_response), 
+                error_message=None
+            )
         except OutputParserException as output_parser_exception:
-            return None, OutputParserException(
-                f"An OutputParserException has occurred for "
-                f"The response from model: {model_response_data.response}\n"
-                f"The exception message: {output_parser_exception}"
+            return ResponseWithError(
+                response=None, 
+                error_message=f"""
+                    An OutputParserException has occurred for the model response: {raw_response}
+                    The exception message: {output_parser_exception}
+                    """
             )
 
-    def _parse_model_output(self, model_responses_data: typing.List[ResponseData]) -> typing.List[ResponseData]:
+    def _parse_model_output(
```

Comment:
This function is called in the `generate` method, which is called in `async _generate` where we have this: https://github.com/allegro/allms/blob/response-json-extraction/allms/models/abstract.py#L179 which is taking the output_dataclass and attaching it to the prompt, like this:

```
self._parser = PydanticOutputParser(pydantic_object=output_data_model_class)
            prompt_template_args[PromptConstants.PARTIAL_VARIABLES_STR] = {
                PromptConstants.OUTPUT_DATA_MODEL: self._parser.get_format_instructions(),
            }
            prompt_template_args[PromptConstants.TEMPLATE_STR] = self._add_output_data_format(prompt=prompt)
```

This creates problems: it's already proven that such prompt will confuse the less capable LLMs.
This is tricky, basically we have the following scenarios:
* the user is using GPT/Gemini/PaLM and passes the output_dataclass --> prompt modified --> successful response --> response parsed success
* the user is using Llama/Mistral/Gemma and passes the output_dataclass --> prompt modified --> unsuccesful response --> response parsed failed

I'm not sure how we can proceed: we probably need to detach the two functionalities, i.e parsing the output via regex and modifying the prompt to get structured data out of it. Maybe, afterall, we need an additional argument provided by the user that will tell the model whether to modify/not modify the prompt.

Reply:
The format instruction injection has been turned off for open-source models.